### PR TITLE
Fix `DataFrame::with_column` to handle creating column names with a period

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

fixes https://github.com/apache/arrow-datafusion/issues/3682

 # Rationale for this change

# What changes are included in this PR?
1. Changes: do not use `col()`, which tries to parse strings as identifiers (e.g. splits on `.`)

# Are there any user-facing changes?
One fewer bug

### CODE REVIEW

File: datafusion/core/src/dataframe.rs [hunk 1]
```diff
@@ -1378,6 +1383,43 @@ mod tests {
         Ok(())
     }
 
+    #[tokio::test]
+    async fn with_column_name() -> Result<()> {
```
Comment (alamb) [inline datafusion/core/src/dataframe.rs:1387]: Thanks to @hengfeiyang for this reproducer

File: datafusion/core/src/dataframe.rs [hunk 2]
```diff
@@ -1378,6 +1383,43 @@ mod tests {
...
+        let ctx = SessionContext::new();
+        ctx.register_batch("t", batch).unwrap();
+
+        let df = ctx
+            .table("t")
+            .unwrap()
+            // try and create a column with a '.' in it
+            .with_column("f.c2", lit("hello"))
+            .unwrap();
+        // Note trying to select causes an error (todo file a separate ticket)
```
Comment (alamb) [inline datafusion/core/src/dataframe.rs:1402]: it turns it something still isn't right (select doesn't work) -- I plan to file an issue (or maybe someone else will have the time to debug it)

**Review by hengfeiyang — APPROVED** [review]

File: datafusion/core/src/dataframe.rs (hunk 2 above)
Comment (alamb) [reply datafusion/core/src/dataframe.rs:1402]: https://github.com/apache/arrow-datafusion/issues/3733

**Comment by hengfeiyang** [comment]
@alamb i tested with #3733 , it looks `with_column` still can't work.

i tested use config:

Cargo.toml
```
datafusion = { git = "https://github.com/apache/arrow-datafusion" }
```

with the example code:

```
use std::sync::Arc;

use datafusion::arrow::array::Int32Array;
use datafusion::arrow::datatypes::{DataType, Field, Schema};
use datafusion::arrow::record_batch::RecordBatch;
use datafusion::datasource::MemTable;
use datafusion::error::Result;
use datafusion::from_slice::FromSlice;
use datafusion::prelude::{col, lit, SessionContext};

/// This example demonstrates how to use the DataFrame API against in-memory data.
#[tokio::main]
async fn main() -> Result<()> {
    // define a schema.
    let schema = Arc::new(Schema::new(vec![Field::new("f.c", DataType::Int32, false)]));

    // define data.
    let batch = RecordBatch::try_new(
        schema.clone(),
        vec![Arc::new(Int32Array::from_slice([1, 10, 10, 100]))],
    )?;

    // declare a new context. In spark API, this corresponds to a new spark SQLsession
    let ctx = SessionContext::new();

    // declare a table in memory. In spark API, this corresponds to createDataFrame(...).
    let provider = MemTable::try_new(schema.clone(), vec![vec![batch]])?;
    ctx.register_table("t", Arc::new(provider))?;
    let df = ctx.table("t")?;

    // construct an expression corresponding to "SELECT * FROM t WHERE f.c = 10" in SQL
    let filter = col("f.c").eq(lit(10));

    let df = df.filter(filter)?;

    // print the results
    df.show().await?;

    Ok(())
}
```

Result:
```
Error: SchemaError(FieldNotFound { qualifier: Some("f"), name: "c", valid_fields: Some(["t.f.c"]) })
```

**Comment by alamb** [comment]
> col("f.c").eq(lit(10));

You can't use `col` with a qualified name (with a `'.'`)

Does it work if you use 

```rust
 Expr::Column(Column {
                        relation: None,
                        name: "f.c".into(),
                    })
```

instead of `col("f.c:)`

?


If so we should probably offer an alternate to `col` that doesn't try and parse the relation that is not as wordy 🤔

**Comment by hengfeiyang** [comment]
Sorry for that, i updated my test code above, I use `with_column`, this case still can't work.

**Comment by alamb** [comment]
> 
Sorry for that, i updated my test code above, I use with_column, this case still can't work.


I will merge this PR and then try your example

**Comment by alamb** [comment]
@hengfeiyang I have merged this PR and now when I run your example from https://github.com/apache/arrow-datafusion/pull/3700#issuecomment-1271752120 it produces this output:

```
+-----+-----------+
| f.c | new_field |
+-----+-----------+
| 1   | hi        |
| 10  | hi        |
| 10  | hi        |
| 100 | hi        |
+-----+-----------+
```

**Comment by hengfeiyang** [comment]
Thanks

**Pull request merged**